packages/react-components/react-dialog/src/components/Dialog/useDialog.ts

```diff
@@ -9,20 +15,181 @@ import type { DialogProps, DialogState } from './Dialog.types';
+      ...overlayShorthand,
+      onClick: handleOverLayClick,
     },
-    // TODO add appropriate slots, for example:
-    // mySlot: resolveShorthand(props.mySlot),
-    root: getNativeElementProps('div', {
-      ref,
-      ...props,
-    }),
+    open,
+    modalType,
+    content,
+    trigger,
+    triggerRef,
+    contentRef,
+    requestOpenChange,
   };
 };
+
+/**
+ * Extracts trigger and content from children
+ */
+function childrenToTriggerAndContent(
+  children: React.ReactNode,
+): readonly [trigger: React.ReactNode, content: React.ReactNode] {
+  const childrenArray = React.Children.toArray(children) as React.ReactElement[];
+  if (process.env.NODE_ENV !== 'production') {
+    if (childrenArray.length !== 1 && childrenArray.length !== 2) {
+      // eslint-disable-next-line no-console
+      console.warn(
+        'Dialog must contain at least one child <DialogContent/>,\n' +
+          'and at most two children <DialogTrigger/> <DialogContent/> (in this order)',
+      );
+    }
+  }
+  switch (childrenArray.length) {
+    // case where there's a trigger followed by content
+    case 2:
+      return childrenArray as [trigger: React.ReactNode, content: React.ReactNode];
+    // case where there's only content
+    case 1:
+      return [undefined, childrenArray[0]];
+    // unknown case
+    default:
+      return [undefined, undefined];
+  }
+}
+
+/**
+ * Checks is click event is a proper Overlay click dismiss
+ */
+function isOverlayClickDismiss(event: React.MouseEvent, type: DialogModalType): boolean {
+  const isDefaultPrevented = normalizeDefaultPrevented(event);
+  return type === 'modal' && !isDefaultPrevented();
+}
+
+/**
+ * Checks if dialog is opening
+ */
+function isOpening(current: boolean, next: Extract<React.SetStateAction<boolean>, Function>) {
+  return !current && next(current) === true;
+}
+
+/**
+ * Focus first element on content when dialog is opened,
+ * in case there's no focusable element, then a eventlistener is added to document
+ * to ensure Escape keydown functionality
+ */
+function useFocusFirstElement({
+  open,
+  requestOpenChange,
+  modalType,
+}: Pick<DialogState, 'open' | 'requestOpenChange' | 'modalType'>) {
+  const { findFirstFocusable } = useFocusFinders();
+  const { targetDocument } = useFluent_unstable();
+  const contentRef = React.useRef<HTMLElement>(null);
+  const triggerRef = React.useRef<HTMLElement>(null);
+
+  React.useEffect(() => {
+    if (!open) {
+      return;
+    }
+
+    const element = contentRef.current && findFirstFocusable(contentRef.current);
+    if (element) {
+      element.focus();
+      // NOTE: if it's non-modal global listener to escape is necessary
+      if (modalType !== 'non-modal') {
+        return;
+      }
+    } else {
+      if (process.env.NODE_ENV !== 'production') {
+        // eslint-disable-next-line no-console
+        console.warn('A Dialog should have at least one focusable element inside DialogContent');
```

Comment:
nit: But I would think on how to emit this warning once to avoid spam in console